rust/noosphere-core/src/bench.rs

```diff
@@ -0,0 +1,182 @@
+use anyhow::Result;
+use instant::Instant;
+use std::{borrow::Cow, future::Future, pin::Pin, time::Duration};
+
+#[cfg(not(target_arch = "wasm32"))]
+type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;
+#[cfg(target_arch = "wasm32")]
+type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + 'a>>;
+
+pub struct BenchDesc {
+    pub name: Cow<'static, str>,
+    pub func: Box<dyn Fn(&mut Bencher) -> BoxFuture<()>>,
+    pub iterations: u32,
+    pub ignore: bool,
+}
+
+#[derive(Default, Debug)]
+pub struct Bencher {
+    start_time: Option<Instant>,
+    pub(crate) duration: Duration,
+    pub bytes: u64,
+}
+
+impl Bencher {
+    pub fn start(&mut self) {
+        if self.start_time.is_some() {
+            panic!("Already started bench.");
+        }
+        self.start_time = Some(Instant::now());
+    }
+
+    pub fn stop(&mut self) {
+        if let Some(start_time) = self.start_time {
+            self.duration = start_time.elapsed();
+        } else {
+            panic!("Bench not started.");
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct BenchStats {
+    iterations: u32,
+    median_duration_ms: u64,
+    median_bytes: u64,
+    name: Cow<'static, str>,
+}
+
+impl From<(BenchDesc, Vec<Bencher>)> for BenchStats {
+    fn from(value: (BenchDesc, Vec<Bencher>)) -> Self {
+        let name = value.0.name;
+        let benchers = value.1;
+        let iterations = benchers.len() as u32;
+        let mut total_bytes = 0;
+        let mut total_duration = 0;
+        for bencher in benchers {
+            total_bytes += bencher.bytes;
+            total_duration += TryInto::<u64>::try_into(bencher.duration.as_millis()).unwrap();
+        }
+
+        BenchStats {
+            name,
+            median_duration_ms: total_duration / iterations as u64,
+            median_bytes: total_bytes / iterations as u64,
+            iterations,
+        }
+    }
+}
+
+fn pad_string(base: &str, column_len: usize, front_pad: bool) -> String {
```

Comment:
I think this might be covered by Rust format strings: https://doc.rust-lang.org/std/fmt/#fillalignment